File: tests/geo2vec_large_scale/evaluate_geo2vec_embeddings.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(arrow)
  library(data.table)
  library(sf)
  library(ggplot2)
})

args <- commandArgs(trailingOnly = TRUE)
get_arg <- function(name, default = NULL) {
  flag <- paste0("--", name)
  idx <- match(flag, args)
  if (is.na(idx)) return(default)
  args[[idx + 1]]
}
has_flag <- function(name) paste0("--", name) %in% args

shape_dir <- get_arg("shape-embedding-dir")
location_dir <- get_arg("location-embedding-dir")
full_dir <- get_arg("full-embedding-dir")
geometry_path <- get_arg("geometry", "/members/dhnyu/fusedatalarge/processed/gwanak_buildings_vworld.gpkg")
layer <- get_arg("layer", "gwanak_buildings")
split_path <- get_arg("split-path", "/members/dhnyu/fusedata/geo2vec_large_scale/metadata/evaluation_splits/gwanak_building_evaluation_split.parquet")
output_dir <- get_arg("output-dir")
seed <- as.integer(get_arg("seed", "20260610"))
max_model_rows <- as.integer(get_arg("max-model-rows", "0"))
xgb_rounds <- as.integer(get_arg("xgb-rounds", "120"))
nthreads_xgboost <- as.integer(get_arg("nthreads-xgboost", "32"))
nthreads_umap <- as.integer(get_arg("nthreads-umap", "32"))
overwrite <- has_flag("overwrite")

if (is.null(shape_dir) || is.null(location_dir) || is.null(full_dir) || is.null(output_dir)) {
  stop("Required: --shape-embedding-dir, --location-embedding-dir, --full-embedding-dir, --output-dir")
}
dir.create(output_dir, recursive = TRUE, showWarnings = FALSE)
manifest_path <- file.path(output_dir, "r_evaluation_manifest.json")
if (file.exists(manifest_path) && !overwrite) {
  cat(readLines(manifest_path), sep = "\n")
  quit(status = 0)
}

set.seed(seed)
pkg_available <- function(pkg) requireNamespace(pkg, quietly = TRUE)
optional_packages <- list(
  ranger = pkg_available("ranger"),
  xgboost = pkg_available("xgboost"),
  uwot = pkg_available("uwot"),
  glmnet = pkg_available("glmnet"),
  FNN = pkg_available("FNN")
)

read_json_simple <- function(path) {
  txt <- paste(readLines(path, warn = FALSE), collapse = "\n")
  jsonlite::fromJSON(txt, simplifyVector = FALSE)
}

read_embedding_dir <- function(path) {
  parts_path <- file.path(path, "embedding_export_parts.parquet")
  parts <- as.data.table(read_parquet(parts_path))
  parts <- parts[order(part_index)]
  rbindlist(lapply(parts$path, function(p) as.data.table(read_parquet(p))), use.names = TRUE)
}

embedding_cols <- function(dt, prefix = NULL) {
  cols <- grep("^geo2vec_", names(dt), value = TRUE)
  if (!is.null(prefix)) cols <- grep(paste0("^", prefix), cols, value = TRUE)
  sort(cols)
}

ring_vertex_count <- function(g) {
  if (inherits(g, "POLYGON")) {
    return(sum(vapply(g, function(r) max(0L, nrow(r) - 1L), integer(1))))
  }
  if (inherits(g, "MULTIPOLYGON")) {
    return(sum(vapply(g, function(poly) sum(vapply(poly, function(r) max(0L, nrow(r) - 1L), integer(1))), integer(1))))
  }
  0L
}

build_labels <- function(building_ids) {
  g <- st_read(geometry_path, layer = layer, quiet = TRUE)
  g$building_id <- as.character(g$building_id)
  g <- g[match(building_ids, g$building_id), ]
  if (any(is.na(g$building_id))) stop("Some embedding building_id values were not found in geometry.")
  geom <- st_geometry(g)
  area <- as.numeric(st_area(geom))
  perimeter <- as.numeric(st_length(st_boundary(geom)))
  bbox <- st_bbox(geom)
  per_geom_bbox <- st_as_sfc(st_bbox(g[1, ]))
  bmat <- do.call(rbind, lapply(seq_along(geom), function(i) as.numeric(st_bbox(geom[i]))))
  bw <- bmat[, 3] - bmat[, 1]
  bh <- bmat[, 4] - bmat[, 2]
  centroid <- st_coordinates(st_centroid(geom))
  vertex_count <- vapply(geom, ring_vertex_count, integer(1))
  data.table(
    building_id = building_ids,
    area = area,
    perimeter = perimeter,
    compactness = ifelse(perimeter > 0, 4 * pi * area / (perimeter^2), NA_real_),
    bbox_aspect_ratio = ifelse(pmin(bw, bh) > 0, pmax(bw, bh) / pmin(bw, bh), NA_real_),
    edge_count = vertex_count,
    vertex_count = vertex_count,
    centroid_x = centroid[, 1],
    centroid_y = centroid[, 2]
  )
}

model_matrix_for <- function(dt, cols) as.matrix(dt[, ..cols])
targets <- c("area", "perimeter", "compactness", "bbox_aspect_ratio", "edge_count", "vertex_count", "centroid_x", "centroid_y")

metric_row <- function(embedding, split_name, target, model, y, pred, n_train, n_test) {
  ok <- is.finite(y) & is.finite(pred)
  y <- y[ok]; pred <- pred[ok]
  ss_res <- sum((y - pred)^2)
  ss_tot <- sum((y - mean(y))^2)
  data.table(
    embedding = embedding,
    split = split_name,
    target = target,
    model = model,
    r2 = ifelse(ss_tot > 0, 1 - ss_res / ss_tot, NA_real_),
    mae = mean(abs(y - pred)),
    train_rows = n_train,
    test_rows = n_test
  )
}

fit_eval <- function(name, dt, cols, split_col) {
  split_name <- sub("^split_", "", split_col)
  if (max_model_rows > 0L && nrow(dt) > max_model_rows) {
    set.seed(seed + nchar(name) + nchar(split_col))
    train_rows <- which(dt[[split_col]] == "train")
    test_rows <- which(dt[[split_col]] == "test")
    train_n <- min(length(train_rows), floor(max_model_rows * 0.8))
    test_n <- min(length(test_rows), max_model_rows - train_n)
    keep <- sort(c(sample(train_rows, train_n), sample(test_rows, test_n)))
    dt <- dt[keep]
  }
  train_idx <- dt[[split_col]] == "train"
  test_idx <- dt[[split_col]] == "test"
  x_train <- model_matrix_for(dt[train_idx], cols)
  x_test <- model_matrix_for(dt[test_idx], cols)
  rows <- list()
  for (target in targets) {
    y_train <- dt[[target]][train_idx]
    y_test <- dt[[target]][test_idx]
    train_df <- data.frame(y = y_train, x_train)
    test_df <- data.frame(x_test)
    lin <- lm(y ~ ., data = train_df)
    rows[[length(rows) + 1L]] <- metric_row(name, split_name, target, "linear_regression", y_test, predict(lin, test_df), sum(train_idx), sum(test_idx))
    if (optional_packages$glmnet) {
      ridge <- glmnet::cv.glmnet(x_train, y_train, alpha = 0, nfolds = 5)
      pred <- as.numeric(predict(ridge, x_test, s = "lambda.min"))
      rows[[length(rows) + 1L]] <- metric_row(name, split_name, target, "ridge_glmnet", y_test, pred, sum(train_idx), sum(test_idx))
    }
    if (optional_packages$ranger) {
      rf <- ranger::ranger(y ~ ., data = train_df, num.trees = 200, max.depth = 12, min.node.size = 3, seed = seed)
      pred <- predict(rf, test_df)$predictions
      rows[[length(rows) + 1L]] <- metric_row(name, split_name, target, "ranger_random_forest", y_test, pred, sum(train_idx), sum(test_idx))
    }
    if (optional_packages$xgboost) {
      dtrain <- xgboost::xgb.DMatrix(x_train, label = y_train)
      dtest <- xgboost::xgb.DMatrix(x_test)
      xgb <- xgboost::xgb.train(
        params = list(
          objective = "reg:squarederror",
          eta = 0.05,
          max_depth = 4,
          subsample = 0.8,
          colsample_bytree = 0.8,
          nthread = nthreads_xgboost
        ),
        data = dtrain,
        nrounds = xgb_rounds,
        verbose = 0
      )
      rows[[length(rows) + 1L]] <- metric_row(name, split_name, target, "xgboost", y_test, predict(xgb, dtest), sum(train_idx), sum(test_idx))
    }
  }
  rbindlist(rows)
}

write_plot <- function(coords, name, method, color_col) {
  fig <- file.path(output_dir, sprintf("%s_%s_%s.png", name, tolower(method), color_col))
  p <- ggplot(coords, aes(x = dim1, y = dim2, color = .data[[color_col]])) +
    geom_point(size = 0.4, alpha = 0.55) +
    scale_color_viridis_c() +
    theme_minimal(base_size = 11) +
    labs(x = paste0(method, "1"), y = paste0(method, "2"), color = color_col)
  ggsave(fig, p, width = 7, height = 5, dpi = 180)
  fig
}

dimension_outputs <- function(name, dt, cols) {
  x <- scale(model_matrix_for(dt, cols))
  pca <- prcomp(x, rank. = 2)
  pca_dt <- data.table(building_id = dt$building_id, dim1 = pca$x[, 1], dim2 = pca$x[, 2], area = dt$area, centroid_x = dt$centroid_x, centroid_y = dt$centroid_y)
  pca_path <- file.path(output_dir, paste0(name, "_pca_coordinates.parquet"))
  write_parquet(pca_dt, pca_path, compression = "zstd")
  out <- list(pca_coordinates = pca_path, pca_area_figure = write_plot(pca_dt, name, "PCA", "area"))
  if (optional_packages$uwot) {
    n <- min(nrow(x), 10000L)
    idx <- sort(sample(seq_len(nrow(x)), n))
    um <- uwot::umap(x[idx, , drop = FALSE], n_neighbors = 30, min_dist = 0.1, n_threads = nthreads_umap, verbose = FALSE)
    um_dt <- data.table(building_id = dt$building_id[idx], dim1 = um[, 1], dim2 = um[, 2], area = dt$area[idx], centroid_x = dt$centroid_x[idx], centroid_y = dt$centroid_y[idx])
    umap_path <- file.path(output_dir, paste0(name, "_umap_coordinates.parquet"))
    write_parquet(um_dt, umap_path, compression = "zstd")
    out$umap_available <- TRUE
    out$umap_coordinates <- umap_path
    out$umap_area_figure <- write_plot(um_dt, name, "UMAP", "area")
  } else {
    out$umap_available <- FALSE
  }
  out
}

retrieval_outputs <- function(name, dt, cols) {
  if (!optional_packages$FNN) return(list(available = FALSE, reason = "FNN unavailable"))
  set.seed(seed)
  x <- scale(model_matrix_for(dt, cols))
  qidx <- sort(sample(seq_len(nrow(dt)), min(25L, nrow(dt))))
  nn <- FNN::get.knnx(x, x[qidx, , drop = FALSE], k = 6)
  rows <- rbindlist(lapply(seq_along(qidx), function(i) {
    qi <- qidx[[i]]
    ni <- nn$nn.index[i, ]
    data.table(
      embedding = name,
      query_building_id = dt$building_id[qi],
      neighbor_rank = seq_along(ni) - 1L,
      neighbor_building_id = dt$building_id[ni],
      embedding_distance = nn$nn.dist[i, ],
      query_area = dt$area[qi],
      neighbor_area = dt$area[ni],
      abs_log_area_delta = abs(log1p(dt$area[qi]) - log1p(dt$area[ni])),
      centroid_distance = sqrt((dt$centroid_x[qi] - dt$centroid_x[ni])^2 + (dt$centroid_y[qi] - dt$centroid_y[ni])^2)
    )
  }))
  path <- file.path(output_dir, paste0(name, "_retrieval_neighbors.parquet"))
  write_parquet(rows, path, compression = "zstd")
  list(available = TRUE, neighbors = path)
}

shape <- read_embedding_dir(shape_dir)
location <- read_embedding_dir(location_dir)
full <- read_embedding_dir(full_dir)
setkey(shape, building_id, geo2vec_internal_id)
setkey(location, building_id, geo2vec_internal_id)
setkey(full, building_id, geo2vec_internal_id)

base <- shape[, .(building_id, geo2vec_internal_id)]
labels <- build_labels(base$building_id)
splits <- as.data.table(read_parquet(split_path))
split_cols <- splits[, .(building_id, split_random, fold_random, split_spatial, fold_spatial)]
dt_base <- Reduce(function(x, y) merge(x, y, by = "building_id"), list(base, labels, split_cols))
setorder(dt_base, geo2vec_internal_id)
missing_targets <- setdiff(targets, names(dt_base))
if (length(missing_targets) > 0) {
  stop(sprintf("Missing evaluation target columns after split join: %s", paste(missing_targets, collapse = ", ")))
}

sets <- list(
  shape = merge(dt_base, shape, by = c("building_id", "geo2vec_internal_id")),
  location = merge(dt_base, location, by = c("building_id", "geo2vec_internal_id")),
  full_geo2vec = merge(dt_base, full, by = c("building_id", "geo2vec_internal_id"))
)
cols <- list(
  shape = embedding_cols(sets$shape, "geo2vec_shp"),
  location = embedding_cols(sets$location, "geo2vec_loc"),
  full_geo2vec = embedding_cols(sets$full_geo2vec)
)

metrics <- rbindlist(unlist(lapply(names(sets), function(nm) {
  list(
    fit_eval(nm, sets[[nm]], cols[[nm]], "split_random"),
    fit_eval(nm, sets[[nm]], cols[[nm]], "split_spatial")
  )
}), recursive = FALSE))
metrics_path <- file.path(output_dir, "r_recoverability_metrics.parquet")
write_parquet(metrics, metrics_path, compression = "zstd")
labels_path <- file.path(output_dir, "r_evaluation_proxy_labels.parquet")
write_parquet(dt_base[, c("building_id", targets, "split_random", "fold_random", "split_spatial", "fold_spatial"), with = FALSE], labels_path, compression = "zstd")

dimensions <- lapply(names(sets), function(nm) dimension_outputs(nm, sets[[nm]], cols[[nm]]))
names(dimensions) <- names(sets)
retrieval <- lapply(names(sets), function(nm) retrieval_outputs(nm, sets[[nm]], cols[[nm]]))
names(retrieval) <- names(sets)

manifest <- list(
  script = "evaluate_geo2vec_embeddings.R",
  complete = TRUE,
  output_dir = output_dir,
  shape_embedding_dir = shape_dir,
  location_embedding_dir = location_dir,
  full_embedding_dir = full_dir,
  split_path = split_path,
  row_count = nrow(dt_base),
  targets = targets,
  optional_packages = optional_packages,
  max_model_rows = max_model_rows,
  xgb_rounds = xgb_rounds,
  nthreads_xgboost = nthreads_xgboost,
  nthreads_umap = nthreads_umap,
  preferred_evaluation_path = "R",
  python_xgboost_umap_required = FALSE,
  geometry_proxies_in_embedding = FALSE,
  recoverability_metrics = metrics_path,
  evaluation_proxy_labels = labels_path,
  dimensions = dimensions,
  retrieval = retrieval
)
jsonlite::write_json(manifest, manifest_path, pretty = TRUE, auto_unbox = TRUE)
cat(jsonlite::toJSON(manifest, pretty = TRUE, auto_unbox = TRUE))

File: tests/geo2vec_large_scale/evaluate_geo2vec_embeddings.py
#!/usr/bin/env python3
"""Evaluate shape, location, and full Geo2Vec embeddings with proxy labels."""

from __future__ import annotations

import argparse
import json
import math
import time
from pathlib import Path
from typing import Any

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq
import pyogrio
from sklearn.decomposition import PCA
from sklearn.ensemble import RandomForestRegressor
from sklearn.linear_model import LinearRegression, Ridge
from sklearn.metrics import mean_absolute_error, r2_score
from sklearn.model_selection import train_test_split
from sklearn.neighbors import NearestNeighbors
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler

from geo2vec_large_scale_common import current_rss_mb, maxrss_mb, path_size_mb, seed_everything, write_json_atomic


TARGETS = [
    "area",
    "perimeter",
    "compactness",
    "bbox_aspect_ratio",
    "edge_count",
    "vertex_count",
    "centroid_x",
    "centroid_y",
]
SHAPE_PROXY_COLUMNS = ["area", "perimeter", "compactness", "bbox_aspect_ratio", "edge_count", "vertex_count"]
LOCATION_PROXY_COLUMNS = ["centroid_x", "centroid_y"]


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--shape-embedding-dir", type=Path, required=True)
    parser.add_argument("--location-embedding-dir", type=Path, required=True)
    parser.add_argument("--full-embedding-dir", type=Path, required=True)
    parser.add_argument("--geometry", type=Path, required=True)
    parser.add_argument("--layer", required=True)
    parser.add_argument("--output-dir", type=Path, required=True)
    parser.add_argument("--seed", type=int, default=20260610)
    parser.add_argument("--test-size", type=float, default=0.25)
    parser.add_argument("--rf-trees", type=int, default=200)
    parser.add_argument("--retrieval-sample", type=int, default=25)
    parser.add_argument("--max-umap-rows", type=int, default=10000)
    parser.add_argument("--overwrite", action="store_true")
    return parser.parse_args()


def read_embedding_dir(path: Path) -> pd.DataFrame:
    parts_path = path / "embedding_export_parts.parquet"
    if not parts_path.exists():
        raise RuntimeError(f"Missing embedding parts manifest: {parts_path}")
    parts = pd.read_parquet(parts_path).sort_values("part_index")
    frames = [pd.read_parquet(p) for p in parts["path"]]
    if not frames:
        raise RuntimeError(f"No embedding parts found under {path}")
    df = pd.concat(frames, ignore_index=True)
    if not df["building_id"].is_unique:
        raise RuntimeError(f"Embedding directory has duplicate building_id values: {path}")
    return df


def embedding_columns(df: pd.DataFrame, prefix: str | None = None) -> list[str]:
    cols = [c for c in df.columns if c.startswith("geo2vec_")]
    if prefix is not None:
        cols = [c for c in cols if c.startswith(prefix)]
    if not cols:
        raise RuntimeError(f"No embedding columns found for prefix={prefix!r}")
    return sorted(cols)


def ring_vertex_count(geom: Any) -> int:
    if geom is None or geom.is_empty:
        return 0
    if geom.geom_type == "Polygon":
        rings = [geom.exterior, *geom.interiors]
        return int(sum(max(0, len(r.coords) - 1) for r in rings))
    if geom.geom_type == "MultiPolygon":
        return int(sum(ring_vertex_count(g) for g in geom.geoms))
    if geom.geom_type == "GeometryCollection":
        return int(sum(ring_vertex_count(g) for g in geom.geoms))
    return 0


def edge_count(geom: Any) -> int:
    return ring_vertex_count(geom)


def build_proxy_labels(geometry: Path, layer: str, building_ids: pd.Series) -> pd.DataFrame:
    gdf = pyogrio.read_dataframe(geometry, layer=layer, columns=["building_id"])
    gdf["building_id"] = gdf["building_id"].astype(str)
    wanted = pd.DataFrame({"building_id": building_ids.astype(str)})
    gdf = wanted.merge(gdf, on="building_id", how="left", validate="one_to_one")
    if gdf["geometry"].isna().any():
        raise RuntimeError("Some embedding building_ids were not found in the geometry layer.")
    geom = gdf["geometry"]
    area = pd.Series([float(g.area) for g in geom], dtype="float64")
    perimeter = pd.Series([float(g.length) for g in geom], dtype="float64")
    compactness = np.where(perimeter > 0, 4.0 * math.pi * area / np.square(perimeter), np.nan)
    bounds = pd.DataFrame([g.bounds for g in geom], columns=["minx", "miny", "maxx", "maxy"])
    bbox_w = (bounds["maxx"] - bounds["minx"]).astype(float)
    bbox_h = (bounds["maxy"] - bounds["miny"]).astype(float)
    short = np.minimum(bbox_w, bbox_h)
    long = np.maximum(bbox_w, bbox_h)
    centroid_x = [float(g.centroid.x) for g in geom]
    centroid_y = [float(g.centroid.y) for g in geom]
    labels = pd.DataFrame(
        {
            "building_id": gdf["building_id"],
            "area": area,
            "perimeter": perimeter,
            "compactness": compactness,
            "bbox_aspect_ratio": np.where(short > 0, long / short, np.nan),
            "edge_count": [edge_count(g) for g in geom],
            "vertex_count": [ring_vertex_count(g) for g in geom],
            "centroid_x": centroid_x,
            "centroid_y": centroid_y,
        }
    )
    labels = labels.replace([np.inf, -np.inf], np.nan)
    if labels[TARGETS].isna().any().any():
        labels = labels.dropna(subset=TARGETS).reset_index(drop=True)
    return labels


def regression_metrics(
    name: str,
    X: np.ndarray,
    labels: pd.DataFrame,
    seed: int,
    test_size: float,
    rf_trees: int,
) -> list[dict[str, Any]]:
    rows: list[dict[str, Any]] = []
    train_idx, test_idx = train_test_split(np.arange(len(labels)), test_size=test_size, random_state=seed)
    models = {
        "linear_regression": make_pipeline(StandardScaler(), LinearRegression()),
        "ridge_regression": make_pipeline(StandardScaler(), Ridge(alpha=1.0)),
        "random_forest": RandomForestRegressor(
            n_estimators=rf_trees,
            max_depth=12,
            min_samples_leaf=3,
            random_state=seed,
            n_jobs=-1,
        ),
    }
    for target in TARGETS:
        y = labels[target].to_numpy(dtype=np.float64)
        for model_name, model in models.items():
            model.fit(X[train_idx], y[train_idx])
            pred = model.predict(X[test_idx])
            rows.append(
                {
                    "embedding": name,
                    "target": target,
                    "model": model_name,
                    "r2": float(r2_score(y[test_idx], pred)),
                    "mae": float(mean_absolute_error(y[test_idx], pred)),
                    "train_rows": int(len(train_idx)),
                    "test_rows": int(len(test_idx)),
                }
            )
    return rows


def pca_outputs(name: str, X: np.ndarray, ids: pd.Series, labels: pd.DataFrame, output_dir: Path) -> dict[str, Any]:
    pca = PCA(n_components=2, random_state=0)
    coords = pca.fit_transform(StandardScaler().fit_transform(X))
    out = pd.DataFrame(
        {
            "building_id": ids.to_numpy(),
            "pc1": coords[:, 0],
            "pc2": coords[:, 1],
            "area": labels["area"].to_numpy(),
            "centroid_x": labels["centroid_x"].to_numpy(),
            "centroid_y": labels["centroid_y"].to_numpy(),
        }
    )
    parquet_path = output_dir / f"{name}_pca_coordinates.parquet"
    pq.write_table(pa.Table.from_pandas(out, preserve_index=False), parquet_path, compression="zstd")
    fig_path = output_dir / f"{name}_pca_area.png"
    plt.figure(figsize=(7, 5))
    sc = plt.scatter(out["pc1"], out["pc2"], c=np.log1p(out["area"]), s=3, alpha=0.55, cmap="viridis")
    plt.colorbar(sc, label="log1p(area)")
    plt.xlabel("PC1")
    plt.ylabel("PC2")
    plt.tight_layout()
    plt.savefig(fig_path, dpi=180)
    plt.close()
    return {
        "coordinates": str(parquet_path),
        "figure": str(fig_path),
        "explained_variance_ratio": [float(x) for x in pca.explained_variance_ratio_],
    }


def umap_outputs(name: str, X: np.ndarray, ids: pd.Series, labels: pd.DataFrame, output_dir: Path, seed: int, max_rows: int) -> dict[str, Any]:
    try:
        import umap  # type: ignore
    except Exception as exc:
        return {
            "available": False,
            "optional": True,
            "error": f"{type(exc).__name__}: {exc}",
            "note": "Python UMAP is optional; R evaluation with uwot is the preferred path when available.",
        }
    if len(X) > max_rows:
        rng = np.random.default_rng(seed)
        idx = np.sort(rng.choice(len(X), size=max_rows, replace=False))
    else:
        idx = np.arange(len(X))
    reducer = umap.UMAP(n_components=2, random_state=seed, n_neighbors=30, min_dist=0.1)
    coords = reducer.fit_transform(StandardScaler().fit_transform(X[idx]))
    out = pd.DataFrame(
        {
            "building_id": ids.iloc[idx].to_numpy(),
            "umap1": coords[:, 0],
            "umap2": coords[:, 1],
            "area": labels["area"].iloc[idx].to_numpy(),
            "centroid_x": labels["centroid_x"].iloc[idx].to_numpy(),
            "centroid_y": labels["centroid_y"].iloc[idx].to_numpy(),
        }
    )
    parquet_path = output_dir / f"{name}_umap_coordinates.parquet"
    pq.write_table(pa.Table.from_pandas(out, preserve_index=False), parquet_path, compression="zstd")
    fig_path = output_dir / f"{name}_umap_area.png"
    plt.figure(figsize=(7, 5))
    sc = plt.scatter(out["umap1"], out["umap2"], c=np.log1p(out["area"]), s=3, alpha=0.55, cmap="viridis")
    plt.colorbar(sc, label="log1p(area)")
    plt.xlabel("UMAP1")
    plt.ylabel("UMAP2")
    plt.tight_layout()
    plt.savefig(fig_path, dpi=180)
    plt.close()
    return {"available": True, "coordinates": str(parquet_path), "figure": str(fig_path), "row_count": int(len(out))}


def retrieval_outputs(name: str, X: np.ndarray, ids: pd.Series, labels: pd.DataFrame, output_dir: Path, seed: int, sample_n: int) -> dict[str, Any]:
    sample_n = min(sample_n, len(X))
    rng = np.random.default_rng(seed)
    query_idx = np.sort(rng.choice(len(X), size=sample_n, replace=False))
    Xs = StandardScaler().fit_transform(X)
    nn = NearestNeighbors(n_neighbors=6, metric="euclidean")
    nn.fit(Xs)
    dist, ind = nn.kneighbors(Xs[query_idx])
    rows = []
    label_lookup = labels.set_index("building_id")
    for qpos, qidx in enumerate(query_idx):
        qid = ids.iloc[qidx]
        qlab = label_lookup.loc[qid]
        for rank, (neighbor_idx, d) in enumerate(zip(ind[qpos], dist[qpos])):
            nid = ids.iloc[int(neighbor_idx)]
            nlab = label_lookup.loc[nid]
            rows.append(
                {
                    "embedding": name,
                    "query_building_id": qid,
                    "neighbor_rank": int(rank),
                    "neighbor_building_id": nid,
                    "embedding_distance": float(d),
                    "query_area": float(qlab["area"]),
                    "neighbor_area": float(nlab["area"]),
                    "abs_log_area_delta": float(abs(np.log1p(qlab["area"]) - np.log1p(nlab["area"]))),
                    "centroid_distance": float(
                        math.hypot(float(qlab["centroid_x"] - nlab["centroid_x"]), float(qlab["centroid_y"] - nlab["centroid_y"]))
                    ),
                }
            )
    out = pd.DataFrame(rows)
    parquet_path = output_dir / f"{name}_retrieval_neighbors.parquet"
    pq.write_table(pa.Table.from_pandas(out, preserve_index=False), parquet_path, compression="zstd")
    summary = (
        out.loc[out["neighbor_rank"] > 0]
        .groupby("embedding", as_index=False)
        .agg(mean_abs_log_area_delta=("abs_log_area_delta", "mean"), mean_centroid_distance=("centroid_distance", "mean"))
    )
    return {"neighbors": str(parquet_path), "summary": summary.to_dict(orient="records")[0]}


def main() -> None:
    args = parse_args()
    manifest_path = args.output_dir / "evaluation_manifest.json"
    if manifest_path.exists() and not args.overwrite:
        print(manifest_path.read_text())
        return
    args.output_dir.mkdir(parents=True, exist_ok=True)
    seed_everything(args.seed)
    start = time.time()

    shape = read_embedding_dir(args.shape_embedding_dir)
    location = read_embedding_dir(args.location_embedding_dir)
    full = read_embedding_dir(args.full_embedding_dir)
    base_ids = shape[["building_id", "geo2vec_internal_id"]].copy()
    for label, df in [("location", location), ("full", full)]:
        if not base_ids.equals(df[["building_id", "geo2vec_internal_id"]].reset_index(drop=True)):
            raise RuntimeError(f"{label} embeddings do not align with shape embeddings.")

    labels = build_proxy_labels(args.geometry, args.layer, base_ids["building_id"])
    merged_ids = base_ids.merge(labels[["building_id"]], on="building_id", how="inner")
    if len(merged_ids) != len(base_ids):
        keep = set(merged_ids["building_id"])
        shape = shape.loc[shape["building_id"].isin(keep)].reset_index(drop=True)
        location = location.loc[location["building_id"].isin(keep)].reset_index(drop=True)
        full = full.loc[full["building_id"].isin(keep)].reset_index(drop=True)
        labels = labels.loc[labels["building_id"].isin(keep)].reset_index(drop=True)

    sets = {
        "shape": (shape, embedding_columns(shape, "geo2vec_shp")),
        "location": (location, embedding_columns(location, "geo2vec_loc")),
        "full_geo2vec": (full, embedding_columns(full)),
    }
    metrics: list[dict[str, Any]] = []
    pca: dict[str, Any] = {}
    umap_info: dict[str, Any] = {}
    retrieval: dict[str, Any] = {}
    dims: dict[str, int] = {}
    for name, (df, cols) in sets.items():
        dims[name] = len(cols)
        X = df[cols].to_numpy(dtype=np.float32)
        ids = df["building_id"].astype(str)
        metrics.extend(regression_metrics(name, X, labels, args.seed, args.test_size, args.rf_trees))
        pca[name] = pca_outputs(name, X, ids, labels, args.output_dir)
        umap_info[name] = umap_outputs(name, X, ids, labels, args.output_dir, args.seed, args.max_umap_rows)
        retrieval[name] = retrieval_outputs(name, X, ids, labels, args.output_dir, args.seed, args.retrieval_sample)

    metrics_df = pd.DataFrame(metrics)
    metrics_path = args.output_dir / "recoverability_metrics.parquet"
    pq.write_table(pa.Table.from_pandas(metrics_df, preserve_index=False), metrics_path, compression="zstd")
    labels_path = args.output_dir / "evaluation_proxy_labels.parquet"
    pq.write_table(pa.Table.from_pandas(labels, preserve_index=False), labels_path, compression="zstd")

    best_rows = (
        metrics_df.sort_values(["target", "model", "r2"], ascending=[True, True, False])
        .groupby(["target", "model"], as_index=False)
        .first()
    )
    summary_table_path = args.output_dir / "recoverability_best_by_target_model.parquet"
    pq.write_table(pa.Table.from_pandas(best_rows, preserve_index=False), summary_table_path, compression="zstd")

    manifest = {
        "script": Path(__file__).name,
        "complete": True,
        "shape_embedding_dir": str(args.shape_embedding_dir),
        "location_embedding_dir": str(args.location_embedding_dir),
        "full_embedding_dir": str(args.full_embedding_dir),
        "geometry": str(args.geometry),
        "layer": args.layer,
        "output_dir": str(args.output_dir),
        "row_count": int(len(labels)),
        "embedding_dimensions": dims,
        "targets": TARGETS,
        "shape_proxy_columns_evaluation_only": SHAPE_PROXY_COLUMNS,
        "location_proxy_columns_evaluation_only": LOCATION_PROXY_COLUMNS,
        "geometry_proxies_in_embedding": False,
        "preferred_evaluation_path": "R",
        "python_xgboost_umap_required": False,
        "optional_python_packages_skipped_are_pipeline_problems": False,
        "recoverability_metrics": str(metrics_path),
        "recoverability_best_by_target_model": str(summary_table_path),
        "evaluation_proxy_labels": str(labels_path),
        "pca": pca,
        "umap": umap_info,
        "retrieval": retrieval,
        "elapsed_seconds": float(time.time() - start),
        "cpu_rss_mb": current_rss_mb(),
        "peak_maxrss_mb": maxrss_mb(),
        "output_size_mb": path_size_mb(args.output_dir),
    }
    write_json_atomic(manifest_path, manifest)
    print(json.dumps(manifest, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()
